In [ ]:
!pip install -q pyomo
from pyomo.environ import *
import numpy as np

     |████████████████████████████████| 9.1 MB 4.1 MB/s 
     |████████████████████████████████| 49 kB 5.7 MB/s 


Question 4.

In [ ]:
coefficient = np.loadtxt('lab4_ex1_coef.txt', delimiter = ',')

In [ ]:
print(coefficient.shape)

(6, 7)


In [ ]:
print('No. of columns=', coefficient.shape[1])

No. of columns= 7


In [ ]:
print('No. of rows=', coefficient.shape[0])

No. of rows= 6


In [ ]:
print(coefficient)

[[  0.85   3.75  -8.75  -3.45 -22.38  20.     0.  ]
 [ -0.7   -3.9   -1.     0.     1.     0.    14.9 ]
 [  1.     0.     0.    -1.     0.    -1.     0.9 ]
 [  0.    -3.    -2.     0.     0.    -1.   -51.7 ]
 [  0.     0.     0.     4.8    9.12   7.2   75.  ]
 [ -0.8   -1.5    8.    -2.     1.     0.   -27.  ]]


Question 5.

In [ ]:
model_ex1 = ConcreteModel()

In [ ]:
#number of decision variables and constraints
M = coefficient.shape[0] - 1 
print(M)

5


In [ ]:
#number of constraints
N = coefficient.shape[1] - 1 
print(N)

6


In [ ]:
#Coefficient of objective function
objective_coefficient = coefficient[0,:-1]

In [ ]:
#constraints coefficient 
constr_coefficient =coefficient[1:,:-1]

In [ ]:
constr_rhs = coefficient[1:,-1]

In [ ]:
#lower bound declaration 
lb = 0

In [ ]:
#upper bound declaration 
ub = np.inf

In [ ]:
# set row indices
row_indices = np.arange(M)
print(row_indices)

[0 1 2 3 4]


In [ ]:
#Set column indices
column_indices = np.arange(N)
print(column_indices)

[0 1 2 3 4 5]


In [ ]:
#decision variables in the model
model_ex1.x = Var(column_indices)

In [ ]:
#constraint list
model_ex1.constraints = ConstraintList()

In [ ]:
for i in row_indices:
     model_ex1.constraints.add(sum(constr_coefficient[i][j]*model_ex1.x[j] for j in column_indices)<= constr_rhs[i])

In [ ]:
for j in column_indices:
  model_ex1.x[j].setlb(lb)
  model_ex1.x[j].setub(ub)

In [ ]:
model_ex1.objective = Objective(expr = summation(objective_coefficient, model_ex1.x), sense = maximize)

In [ ]:
model_ex1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   inf : False :  True :  Reals
          1 :     0 :  None :   inf : False :  True :  Reals
          2 :     0 :  None :   inf : False :  True :  Reals
          3 :     0 :  None :   inf : False :  True :  Reals
          4 :     0 :  None :   inf : False :  True :  Reals
          5 :     0 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 0.85*x[0] + 3.75*x[1] - 8.75*x[2

$$Q-6$$

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 155047 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model_ex1)
print('The Solver status:', result.solver.status)
print('The Solver termination condition:',result.solver.termination_condition)

    model.name="unknown";
      - termination condition: unbounded
      - message from solver: <undefined>
The Solver status: warning
The Solver termination condition: unbounded


Question 7.

In this problem ,the termination condition is unbounded that is we can maximize the objective function indefinitely without violating the constraints. \\
The solver status shows that we cannot find an optimal solution , as the feasible region is not bounded above that is unbounded.

Question 8.

We were solving the problem in the wrongn optimization sense ,so the feasible region is unbounded.

In [ ]:
#the sense reset to minimized
model_ex1.objective.set_sense(minimize)
model_ex1.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   inf : False :  True :  Reals
          1 :     0 :  None :   inf : False :  True :  Reals
          2 :     0 :  None :   inf : False :  True :  Reals
          3 :     0 :  None :   inf : False :  True :  Reals
          4 :     0 :  None :   inf : False :  True :  Reals
          5 :     0 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 0.85*x[0] + 3.75*x[1] - 8.75*x[2

In [ ]:
result = opt_cbc.solve(model_ex1)
print(result)
print('SolverStatus:', result.solver.status)
print('Termination Condition:', result.solver.termination_condition)


Problem: 
- Name: unknown
  Lower bound: -97.02184211
  Upper bound: -97.02184211
  Number of objectives: 1
  Number of constraints: 6
  Number of variables: 7
  Number of nonzeros: 6
  Sense: minimize
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: None
      Number of created subproblems: None
    Black box: 
      Number of iterations: 6
  Error rc: 0
  Time: 0.028368234634399414
Solution: 
- number of solutions: 0
  number of solutions displayed: 0

SolverStatus: ok
Termination Condition: optimal


In [ ]:
# display results
print('Objective Function:', model_ex1.objective())
print('\nDecision Variables: ')
for j in column_indices:
      print('x[',j,']= ', model_ex1.x[j].value)
print('\nConstraints:')
model_ex1.constraints.display()      

Objective Function: -97.02184239599997

Decision Variables: 
x[ 0 ]=  0.9
x[ 1 ]=  23.002456
x[ 2 ]=  0.0
x[ 3 ]=  0.0
x[ 4 ]=  8.2236842
x[ 5 ]=  0.0

Constraints:
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :         -82.1158942 :  14.9
      2 :  None :                 0.9 :   0.9
      3 :  None :          -69.007368 : -51.7
      4 :  None :   74.99999990399999 :  75.0
      5 :  None : -26.999999799999998 : -27.0


Question 9

After the change in sense from maximize to minimize , we are now getting an optimal solution .The previous termination condition was due to wrong sense of the optimization problem . \\
Objective Value = -97.02184 \\
Optimal solution : x1=0.9 ; x2=23.0024,x3=0,x4=0,x5=8.2236842,x6=0 \\
Constraint 2 ,4 and 5 (last 2 in approximate sense) are active since they exact (approximate) equality